src/serializer/serializer.js

```diff
@@ -66,18 +66,32 @@ export class Serializer {
   statistics: SerializerStatistics;
   react: ReactSerializerState;
 
+  _populateForbiddenNames(ast: BabelNode, realmPreludeGenerator: PreludeGenerator): void {
+    let forbiddenNames = realmPreludeGenerator.nameGenerator.forbiddenNames;
+
+    // by finding if an identifier is pure, we know if it's leaked to the global or not
```

Comment:
I don't fully understand the reasoning here. It shouldn't matter if a name "leaks" or not --- if it's used, then it potentially clashes with any name we might generated and place in any scope.

Reply:
This breaks FB www rules around certain naming cases, so there’s interval correctness to this PR too.

Reply:
Note I don't think this would fix the internal naming problem in all cases. It solves the *common* case (where they don't actually clash). But if they do, this won't help.

So this is not blocking my work. This solution won't be a full fix (even if it worked fully).

Reply:
What do you both suggest we do then? Drop this PR and find another solution? Maybe I could split out the traversal work and the work in the generator (keeping that would make for better naming already).

Reply:
For the FB use case I think we should have some preprocessing step that just turns them into globals (and teach Prepack to treat them as globals). Because our transform doesn't actually care if they're imported or not.

Reply:
I guess the problem with *that* approach is that we actually need to leave them as `cx`, not `global.cx` 🤔